### DISCUSSION TITLE
is getCsrfToken() on the client side "safe"?

### DISCUSSION THREAD

**Discussion by eak12913** [nextauthjs/next-auth#1868] [Help] [answered]
I was always under the impression that CSRF token would be generated on the server and "baked into" the HTML that's returned. I see that next-auth provides this API endpoint that gets you a CSRF token and I was worried that this could somehow be less secure or could expose us to a larger number of threat vectors. 

I noticed that the [signIn](https://next-auth.js.org/getting-started/client#signin) helper method makes use of this client side CSRF helper. I'd be grateful if folks chime in and tell me that my concerns are based on false assumptions or incorrect understanding!

**Comment by SchalkBezuidenhout** [comment]
Something that bothered me was that anyone could view the '/api/auth' routes directly, including the '/api/auth/csrf'.

I wrote a simple middleware that takes care of this, by checking the referer header.

`export const validateReferer = fn => async (req, res) => {
    const { referer } = req.headers;
    if (!referer?.includes(process.env.NEXTAUTH_URL)) {
        console.log('validation failed');
        return res.status(403).json()
    } else {
        console.log('successfully passed validation');
        return fn(req, res);
    }
}`

Stick this in the [...nextauth].js export and the routes are only accessible if the referer header includes the NEXTAUTH_URL set in .env.

`export default validateReferer((req, res) => NextAuth(req, res, options));`

**Reply by balazsorban44** [reply]
Strictly speaking, none of the headers are safe either, as I could set whatever I want, can't I?

**Reply by balazsorban44** [reply]
See https://github.com/nextauthjs/next-auth/discussions/1868#discussioncomment-912544 for a better explanation why this might not be needed

**Comment by balazsorban44** [comment]
The getCsrfToken is just the read of a httponly cookie (it IS generated server-side), that cannot be written from client-side. it is used in our signin function, and as long as you use that, I don't see a problem.

There is a chance that we wouldn't even need to attach the csrfToken the way we do it in the signin function, as the cookie will be sent with the `fetch` anyway? 🤔

**Comment by balazsorban44** [comment, marked as answer]
After digging into our code, I found out that we are using a technique called `Double Submit Cookie` described by OWASP here:
https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie
and also here:
https://owasp.org/www-chapter-london/assets/slides/David_Johansson-Double_Defeat_of_Double-Submit_Cookie.pdf

So in our case, the current way seems to be the ONLY secure way, actually. :slightly_smiling_face:

Source code: https://github.com/nextauthjs/next-auth/blob/24ee459f9749387d4e7952dd8d3c9f0f93dd3651/src/server/lib/csrf-token-handler.js#L4-L16

**Comment by advename** [comment]
@balazsorban44  I stumbled upon this issue by chance and after reading through the comments, I noticed that neither the CSRF token nor the cookie is using an encryption technique, as recommended by [OWASP Double cookie submit](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie)

Shouldn't this be implemented in case of a Subdomain takeover attack, or subdomain vulnerability?
